src/native/libs/System.Native/pal_io.c

```diff
@@ -61,12 +61,33 @@ extern int     getpeereid(int, uid_t *__restrict__, gid_t *__restrict__);
 #endif
 #endif
 
-// The portable build is performed on RHEL7 which doesn't define FICLONE yet.
+#ifdef TARGET_LINUX
+#include <sys/utsname.h>
+
 // Ensure FICLONE is defined for all Linux builds.
-#ifdef __linux__
 #ifndef FICLONE
 #define FICLONE _IOW(0x94, 9, int)
 #endif
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wreserved-id-macro"
+// Ensure __NR_copy_file_range is defined for portable builds.
+#include <sys/syscall.h> // __NR_copy_file_range
+# if !defined(__NR_copy_file_range)
+#  if defined(__amd64__)
+#   define __NR_copy_file_range  326
+#  elif defined(__i386__)
+#   define __NR_copy_file_range  377
+#  elif defined(__arm__)
+#   define __NR_copy_file_range  391
+#  elif defined(__aarch64__)
+#   define __NR_copy_file_range  285
+#  else
+#   error Unknown architecture
+#  endif
```

Comment:
@am11 are the wasm compilation failures due to using `TARGET_LINUX` instead of `__linux__`? Or is `__linux__` also defined for the wasm compilation?

Reply:
To exclude wasm, `#if defined(TARGET_LINUX) && !defined(TARGET_WASM)`

Reply:
Is `__linux__` defined for the wasm compilation?

Reply:
Nope, wasm mostly conforms to POSIX.